armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+        raise NotImplementedError("hooking not ready for tensorflow")
+        # NOTE:
+        # https://discuss.pytorch.org/t/get-the-activations-of-the-second-to-last-layer/55629/6
+        # TensorFlow hooks
+        # https://www.tensorflow.org/api_docs/python/tf/estimator/SessionRunHook
+        # https://github.com/tensorflow/tensorflow/issues/33478
+        # https://github.com/tensorflow/tensorflow/issues/33129
+        # https://stackoverflow.com/questions/48966281/get-intermediate-output-from-keras-tensorflow-during-prediction
+        # https://stackoverflow.com/questions/59493222/access-output-of-intermediate-layers-in-tensor-flow-2-0-in-eager-mode/60945216#60945216
+
+    def hook_torch(self, module, *preprocessing, input=None, output=None):
+        if not hasattr(module, "register_forward_hook"):
+            raise ValueError(
+                f"module {module} does not have method 'register_forward_hook'. Is it a torch.nn.Module?"
+            )
+        if input == "" or (input is not None and not isinstance(input, str)):
+            raise ValueError(f"input {input} must be None or a non-empty string")
+        if output == "" or (output is not None and not isinstance(output, str)):
+            raise ValueError(f"output {output} must be None or a non-empty string")
+        if input is None and output is None:
+            raise ValueError("input and output cannot both be None")
+        if module in self._hooks:
+            raise ValueError(f"module {module} is already hooked")
+
+        def hook_fn(hook_module, hook_input, hook_output):
+            del hook_module
+            key_values = {}
+            if input is not None:
+                key_values[input] = hook_input
+            if output is not None:
+                key_values[output] = hook_output
+            self.update(*preprocessing, **key_values)
+
+        hook = module.register_forward_hook(hook_fn)
+        self._hooks[module] = (hook, "pytorch")
+
+    def unhook(self, module):
+        hook, mode = self._hooks.pop(module)
+        if mode == "pytorch":
+            hook.remove()
+        elif mode == "tf":
+            raise NotImplementedError()
+        else:
+            raise ValueError(f"mode {mode} not in ('pytorch', 'tf')")
+
+
+def process_meter_arg(arg: str):
+    """
+    Return the probe variable and stage_filter
+
+    Example strings: 'model.x2[adversarial]', 'scenario.y_pred'
+    """
+    if "[" in arg:
+        if arg.count("[") != 1 and arg.count("]") != 1:
+            raise ValueError(f"arg {arg} must have a single matching [] or none")
+        arg, filt = arg.split("[")
+        if filt[-1] != "]":
+            raise ValueError(f"arg {arg} cannot have chars after final ']'")
+        stage_filter = filt[:-1].strip()
+        # tokens = [x.strip() for x in filt.split(",")]
+    else:
+        stage_filter = None
+    probe_variable = arg
+
+    return probe_variable, stage_filter
+
+
+class ProbeMapper:
```

Comment:
I want to make sure I understand how all the pieces (probe, meter, context, probemapper, writer) fit together. A context has a bunch of meters. Each meter corresponds to a metric you want to measure. A meter uses values stored via `probe.update()` calls to compute a metric

I don't quite see how/where a `ProbeMapper` fits in. Perhaps in other words, what is it doing that the context object do can't do? The `map_to_meters_args()` method and corresponding nested dictionary `self.probe_filter_meter_arg` I'm finding a little difficult to wrap my head around

Reply:
`ProbeMapper` is meant to store mappings from probe outputs to meter inputs. It is not envisioned to be used outside of Context (since context args like "stage" must be passed in). However, since it is essentially a data structure, I separated it into its own class for easier testing.